topics/single-cell/tutorials/scrna-ncbi-anndata/tutorial.md

```diff
@@ -0,0 +1,471 @@
+>    - *"4: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `0|1|7`
+>         - *"Replace with"*: `NA`
+>    - **+ Insert Replacement**
+>    - *"5: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `batch`
+>         - *"Replace with"*: `replicate`
+>
+> 2. {% tool [Cut](Cut1) %} with the following parameters:
+>    - *"Cut columns"*: `c2`
+>    - *"Delimited by"*: `Tab`
+>    - {% icon param-file %} *"From"*: `output of Replace Text`
+>
+> 3. **Rename** {% icon galaxy-pencil %} output `Replicate Metadata`
+>
+{: .hands_on}
+
+Next we will add the metadata indicating which patient each row came from.
+
+> <hands-on-title>Create patient data</hands-on-title>
+>
+> 1. {% tool [Replace Text](toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_replace_in_column/1.1.3) %} with the following parameters:
+>    - {% icon param-file %} *"File to process"*: `Observation data`
+>    - *"1: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `(0$)|(1$)`
+>         - *"Replace with"*: `patient1`
+>    - **+ Insert Replacement**
+>    - *"2: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `(2$)|(3$)|(4$)|(5$)|(6$)`
+>         - *"Replace with"*: `patient2`
+>    - **+ Insert Replacement**
+>    - *"3: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `(7$)|(8$)|(9$)`
+>         - *"Replace with"*: `patient3`
+>    - **+ Insert Replacement**
+>    - *"5: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `batch`
+>         - *"Replace with"*: `patient`
+>
+> 2. {% tool [Cut](Cut1) %} with the following parameters:
+>    - *"Cut columns"*: `c2`
+>    - *"Delimited by"*: `Tab`
+>    - {% icon param-file %} *"From"*: `output of Replace Text`
+>
+> 3. **Rename** {% icon galaxy-pencil %} output `Patient Metadata`
+>
+{: .hands_on}
+
+We will now add a column to indicate which sample each row came from using the sample ID's described earlier.
+
+> <hands-on-title>Create sample id metadata</hands-on-title>
+>
+> 1. {% tool [Replace Text](toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_replace_in_column/1.1.3) %} with the following parameters:
+>    - {% icon param-file %} *"File to process"*: `Observation data`
+>    - *"1: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `0`
+>         - *"Replace with"*: `AUG_PB1A`
+>    - **+ Insert Replacement**
+>    - *"2: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `1$`
+>         - *"Replace with"*: `AUG_PB1B`
+>    - **+ Insert Replacement**
+>    - *"3: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `(2$)|3`
+>         - *"Replace with"*: `MAY_PB1A`
+>    - **+ Insert Replacement**
+>    - *"4: Replacement"*
+>         - *"in column"*: `Column: 2`
+>         - *"Find pattern"*: `4|5|6`
```

Comment:
why sometimes without `$`?

Reply:
It's added to stop earlier replace operations being replaced again (e.g. the 1 in AUG_PB1A will be replaced with "AUG_PB1B" in the second replacement operation) I was meant to make a note of this but forgot, I'll add in an explanation!